### PR TITLE
Basic implementation of sequential operations dispatching

### PR DESCRIPTION
**Pull request by gravikovich**
Based on patterns used in several places in SFINCS

### CODE REVIEW

**Comment by realgpasternak** [comment]
Please allow providing DataflowBlockOptions on construction, so we will be able to use this with ensured order and specify parallelism level (in our use case we use it to force 1 task at a time)

File: Bsii.Dotnet.Utils/Sequential/BufferBlockSequentialDispatcher.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+            public Task<T> OperationRepresentation => _tcs.Task;
+
+
+            public DispatchedOperation(Func<Task<T>> exec)
+            {
+                _exec = exec;
+            }
+
+            public async Task ExecuteAsync()
```
Comment (kpahima) [inline Bsii.Dotnet.Utils/Sequential/BufferBlockSequentialDispatcher.cs:35]: What if dispatching has input it needs?

Comment (gravikovich) [reply Bsii.Dotnet.Utils/Sequential/BufferBlockSequentialDispatcher.cs:35]: It captures anything it needs in the `Actions` it dispatches.

**Comment by gravikovich** [comment]
> Please allow providing DataflowBlockOptions on construction, so we will be able to use this with ensured order and specify parallelism level (in our use case we use it to force 1 task at a time)

Done, but note that I use `BufferBlock` not `ActionBlock`, and handle the operations manually in a single loop, so parallelism level here is 1 by definition. I believe the same is true for order ensuring, though not sure.

File: Bsii.Dotnet.Utils/Sequential/BufferBlockSequentialDispatcher.cs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+            while (await _operations.OutputAvailableAsync())
+            {
+                var op = await _operations.ReceiveAsync();
+                await op.ExecuteAsync(); //This is guaranteed not to throw
+            }
+        }
+
+        public void Start()
+        {
+            if (_isStarted)
```
Comment (eitanhs) [inline Bsii.Dotnet.Utils/Sequential/BufferBlockSequentialDispatcher.cs:123]: Don't we want to be extra cautious and use Interlocked.Exchange...?

File: Bsii.Dotnet.Utils/Sequential/BufferBlockSequentialDispatcher.cs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            {
+                throw new InvalidOperationException("The dispatcher was already started");
+            }
+            _isStarted = true;
+            Task.Run(DispatchOperations);
+        }
+
+        public Task StopAsync()
+        {
+            if (!_isStarted)
```
Comment (eitanhs) [inline Bsii.Dotnet.Utils/Sequential/BufferBlockSequentialDispatcher.cs:133]: Don't we want to be extra cautious and use Interlocked.Exchange...?

File: Bsii.Dotnet.Utils/Sequential/ISequentialOperationsDispatcher.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
+﻿using System;
+using System.Threading.Tasks;
+
+namespace Bsii.Dotnet.Utils.Sequential
+{
+    /// <summary>
+    /// Guarantees sequential execution of operations dispatched to it
+    /// Note - nested dependent operations will cause deadlocks
+    /// </summary>
+    public interface ISequentialOperationsDispatcher
```
Comment (eitanhs) [inline Bsii.Dotnet.Utils/Sequential/ISequentialOperationsDispatcher.cs:10]:
```suggestion
    public interface ISequentialOperationsDispatcher : IAsyncDisposable
```

File: Bsii.Dotnet.Utils/Sequential/ISequentialOperationsDispatcher.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        /// </summary>
+        void Start();
+
+        /// <summary>
+        /// Stops processing of the commands, doesn't accept new commands afterwards
+        /// </summary>
+        /// <returns>the task that can be awaited for notification of operations completion</returns>
+        Task StopAsync();
+
+    }
```
Comment (eitanhs) [inline Bsii.Dotnet.Utils/Sequential/ISequentialOperationsDispatcher.cs:44]: I hope I got it right:
```suggestion
        /// <<inheritdoc/>
        ValueTask DisposeAsync() => new ValueTask(StopAsync());
    }
```

File: Bsii.Dotnet.Utils/Sequential/SequentialOperationsDispatcherExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        /// <summary>
+        /// Executes an action, guarantees no other dispatches run in parallel with it
+        /// Note - The dispatching is synchronous and doesn't perform the execution, the returned task is to represent the result
+        /// Note - Nested dependent operations will cause deadlocks
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="exec"></param>
+        /// <returns>A task that represents the status of the action execution, can be used to get the results</returns>
+        public static Task Dispatch(this ISequentialOperationsDispatcher dispatcher, Action exec)
```
Comment (eitanhs) [inline Bsii.Dotnet.Utils/Sequential/SequentialOperationsDispatcherExtensions.cs:16]: Love it! elegant solution :)

File: Bsii.Dotnet.Utils/Sequential/BufferBlockSequentialDispatcher.cs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        #endregion
+
+        private readonly BufferBlock<IDispatchedOperation> _operations;
+        private bool _isStarted = false;
+
+        /// <summary>
+        /// Creates the dispatcher optionally specifying dataflow block options
+        /// </summary>
+        /// <param name="blockOptions"></param>
+        public BufferBlockSequentialDispatcher(DataflowBlockOptions blockOptions = null)
```
Comment (eitanhs) [inline Bsii.Dotnet.Utils/Sequential/BufferBlockSequentialDispatcher.cs:87]: > Please allow providing DataflowBlockOptions on construction, so we will be able to use this with ensured order and specify parallelism level (in our use case we use it to force 1 task at a time)

As this class name suggest - it is by definition sequential..
so I'm not sure supporting custom `DataflowBlockOptions` doesn't make more confusion than flexibility

File: Bsii.Dotnet.Utils/Sequential/BufferBlockSequentialDispatcher.cs (hunk 2 above)
Comment (gravikovich) [reply Bsii.Dotnet.Utils/Sequential/BufferBlockSequentialDispatcher.cs:123]: Don't think so, `Start/Stop` should not be thread safe, you shouldn't call them concurrently, it's the general convention for all objects. Only `Dispatch` is thread safe.

File: Bsii.Dotnet.Utils/Sequential/BufferBlockSequentialDispatcher.cs (hunk 3 above)
Comment (gravikovich) [reply Bsii.Dotnet.Utils/Sequential/BufferBlockSequentialDispatcher.cs:133]: As above

**Comment by realgpasternak** [comment]
You are both correct, I confused with the options for ActionBlock: https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.dataflow.executiondataflowblockoptions?view=netcore-3.1

Also, actually EnsureOrdered defaults to true, just need to be aware of it.

File: Bsii.Dotnet.Utils/Sequential/ISequentialOperationsDispatcher.cs (hunk 4 above)
Comment (gravikovich) [reply Bsii.Dotnet.Utils/Sequential/ISequentialOperationsDispatcher.cs:10]: Bsii.Utils is .net standard 2.0, no `IAsyncDisposable` there... I will try to implement regular `IDisposable`

File: Bsii.Dotnet.Utils/Sequential/ISequentialOperationsDispatcher.cs (hunk 5 above)
Comment (gravikovich) [reply Bsii.Dotnet.Utils/Sequential/ISequentialOperationsDispatcher.cs:44]: As above

File: Bsii.Dotnet.Utils/Sequential/BufferBlockSequentialDispatcher.cs (hunk 7 above)
Comment (gravikovich) [reply Bsii.Dotnet.Utils/Sequential/BufferBlockSequentialDispatcher.cs:87]: Agreed, removed

**Pull request merged**